### DISCUSSION TITLE
Maximum frequency tracker

### DISCUSSION THREAD

**Discussion by t-anyu** [cyring/CoreFreq#266] [Ideas]
Hey,

CoreFreq is awesome. To make it awesomer, a maximum frequency tracker would be great to see how high the single core (or less than allcore) will reach. Extremely useful in overclocking or just curious if a new CPU can hit the target speeds (or above).

Thanks

**Comment by cyring** [comment]
This is a great idea.
Q: Is the maximum frequency tracker about the highest frequency since _CoreFreq_ start-up ?

If yes, parts are already programmed into the Daemon which stores the frequency limits per Core.

* Relative structure
https://github.com/cyring/CoreFreq/blob/506902641d2097e2caaaf3c71a0248133419f25a/corefreq.h#L245
* Absolute structure
https://github.com/cyring/CoreFreq/blob/506902641d2097e2caaaf3c71a0248133419f25a/corefreq.h#L248

Proposals
1. Limits will be added to the JSON output (easy)
2. UI: space is missing in the main `Frequency` view, constrained to fit in 80 columns (refactoring)
3. Add a new view dedicated to frequency limits ? (better)
4. Other suggestion ?

**Comment by cyring** [comment]
For your testings, frequency limits are available in `develop` branch, commit 20fd57fd0ec6b74d6c5d05c6ab59fa617052ce04

You should get such output:
![CoreFreq_JSON_CPU_Frequency_Limit](https://user-images.githubusercontent.com/11563789/123677394-5bd59880-d845-11eb-9bb0-06647e1018a3.png)

**Reply by t-anyu** [reply]
Hey,

Yes maximum frequency since CoreFreq start-up. Users can typically reset that manually or restart. This is common with HWInfo (on Windows side).

An another view with frequency limits would work also if that's the easiest.

Thanks for the new branch. I'll have a test!

**Reply by cyring** [reply]
> Hey,
> 
> Yes maximum frequency since CoreFreq start-up. Users can typically reset that manually or restart. This is common with HWInfo (on Windows side).
> 
> An another view with frequency limits would work also if that's the easiest.
> 
> Thanks for the new branch. I'll have a test!

Feel free to suggest a view design if some columns matter

Btw, limits (Vcore, Power, Frequency, Temperature) can be reset at anytime in the UI, pressing the star key.

**Reply by t-anyu** [reply]
Hey,

just tested. Perfect I can now see that my new 5950x is hitting 5GHz on some cores and some are not.

I think ideally: If there's a column limit, I would replace ratio or turbo column to maximum frequency. Or preferably ratio, because ratio is typically nowdays always ~ Freq / 100 (even with the spread spectrum FCLK clock generator option enabled), so it doesn't give that much useful information. But this is just my opinion.


0 4999.96255
1 4999.9926
2 4999.9976
3 4999.98865
4 5000.00005
5 4999.99455
6 4999.99555
7 4999.1441
8 4899.987897
9 4899.981086
10 4899.976235
11 4899.989465
12 4899.997599
13 4800.002976
14 4800.01872
15 4799.993472
16 4999.9532
17 4999.95855
18 5000.01285
19 4999.99525
20 5000.0321
21 5000.00135
22 4999.98175
23 4999.9908
24 4899.996227
25 4899.99804
26 4899.991817
27 4900.000735
28 4899.952323
29 4799.991216
30 4799.992608
31 4799.983008

**Reply by cyring** [reply]
Yes, offering a shortcut to toggle between `Ratio` and `Max` frequency

And save last User choice into the `corefreq.cfg`

Give me some time to think about impacts on UI latency (I try to minimize branches as much as possible)

**Comment by cyring** [comment]
Commit 93e850ef53df915d0ee6d4a95a7ea9d5f85e543f of the `develop` branch is proposing a **Custom** view.
* You will press the shortcut [`y`] to switch to this view
* You can also start it from the command line `corefreq-cli -t custom`

![CoreFreq_Custom_View](https://user-images.githubusercontent.com/11563789/123896920-c4666780-d962-11eb-85de-ea2ed680067b.png)

* The Custom view won't fit on standard Consoles `80x25` but those becomes rare nowadays
* The Footer ruler should follow with a mix of average and per Package values
*  Among the last tasks will come choice of colors and optimizations

Thanks for any comment.

**Reply by t-anyu** [reply]
Hey, awesome!

I'll test it out.

**Reply by cyring** [reply]
> Hey, awesome!
> 
> I'll test it out.

* And if space is missing because you have a High-Core-Count processor, think about mixing those build directives:
```
make NO_UPPER=1 NO_HEADER=1 NO_FOOTER=1 NO_LOWER=1
```
![CoreFreq_Right_Aligned_Volt_Temp](https://user-images.githubusercontent.com/11563789/123971578-79783e80-d9ba-11eb-973d-fabc22cd08ee.png)

* The CPUs on the area can scroll up or down using shortcuts `+` or `-`
* The same with `Down` and `Up` keys when no window is opened.

**Comment by cyring** [comment]
Are your guys familiar with `snprintf`, its format, arguments, and so on ?

**Reply by cyring** [reply]
> Are your guys familiar with `snprintf`, its format, arguments, and so on ?

Some customization like this:
```
make FIELD='%7.2f %7.2f %7.2f   %7.2f %7.2f %7.2f  %3u %3u %3u  %7.4f %7.4f %7.4f  %8.4f %8.4f %8.4f  %6.2f %6.2f %6.2f %6.2f %6.2f %6.2f  %10.6f%10.6f%10.6f'
```

Hard to employ ?

**Comment by cyring** [comment]
Hello,

* The Custom view is available for testing since commit b6d308d3df5676f4be689513e84dd3e3fbfeb8c1
![CoreFreq_CustomViewMenu](https://user-images.githubusercontent.com/11563789/124311704-6d34e280-db6e-11eb-82b8-f64b45d743ef.png)

* Two build options `CUSTOM_RULER` and `CUSTOM_FIELD` can be processed by the `Makefile`
  `CUSTOM_FIELD`  follows the `printf` syntax 
  ( If used in the format string, symbol dollar needs to be written twice )
```
make help
```
```
o---------------------------------------------------------------o
|  make [all] [clean] [info] [help] [install] [module-install]  |
...
|                                                               |
|  Custom view ( printf syntax ):                               |
|    CUSTOM_RULER='string'                                      |
|    CUSTOM_FIELD='format'                                      |
|                                                               |
...
o---------------------------------------------------------------o
```
* The _CoreFreq_ Client can also be immediately started with your built Custom view
```
corefreq-cli -t custom
```

**Comment by cyring** [comment]
The attached archive is providing a change for the Field format and the Ruler string as Client arguments.
```
corefreq-cli -h
```
```
...
Usage:  corefreq-cli [-Option <argument>] [-Command <argument>]

    Interface options
...
        -Yf 's' Custom field format
        -Yr 's' Custom ruler string
...
```
It allows User to try various customization without rebuilding the project each time.
```
corefreq-cli -Yf '%5$7.2f %6$7.2f %4$7.2f %3$7.2f %2$7.2f %1$7.2f' -Yr '------ Freq. --- High ------' -t custom
```

Syntax just follows the plain specification of `printf` **but** with the drawback that any wrong specifier can lead to an undefined behavior.

Not sure if I want to pursue with such weak implementation but creating a _control-format_ will cost a lot of coding !

[CoreFreq_develop.tar.gz](https://github.com/cyring/CoreFreq/files/6796141/CoreFreq_develop.tar.gz)

**Comment by cyring** [comment]
That would be the final layout 5adce7f371097768463300e251fe171a105f9552
( _I'm removing any User variables to avoid issues_ )

![CoreFreq_Laout_Custom_view](https://user-images.githubusercontent.com/11563789/125193413-9cbdac00-e24c-11eb-89e8-77732ded7dcb.png)

**Comment by cyring** [comment]
* Latest `develop` branch provides Package values. 
*  Edit: especially Package C-States for Intel gifted processors. 
  The voltage all Cores sum and Power`|`Energy values are more Registers contextual, like RAPL availability.

![CoreFreq_Custom_Pkg](https://user-images.githubusercontent.com/11563789/125764539-c532df16-4adf-4244-9ee0-8faf0ce13043.png)

* In addition to that are fixed SMT issues which claimed **non regression tests on all platforms** = thanks anyone for your confirmation ?

  Edit: rolling back the Driver changes which don't fix the max frequency limit issue.

**Comment by cyring** [comment]
Other Power fields are now added in the footer.

![CoreFreq_Custom_Footer](https://user-images.githubusercontent.com/11563789/126018846-7687a45b-351a-475c-bd78-cedde3a33388.png)

**Comment by cyring** [comment]
@t-anyu

Hello,

Does _CoreFreq_ have an issue to monitor the Turbo Boost frequencies on 5900X / 5950X ?

Can you please test from latest version and show me screenshots single and all Cores boost.

Using the integrated stress tool
